ts/doclets-copy-extensions

KacperMadej opened the pull request:
Despite the linked Asana task, branch name and the name of the PR this won't handle all files in Extensions directory because this is too big to handle at once. The PR is limited to Extensions/A*-E* directories with files inside. Another PR will be created to finish the main task.

Additional changes or comments:
- [x] rename .d.ts to .ts
- partial review for Stock related files:
- [x] Annotations
- [x] DataGrouping
- also see comments in https://github.com/highcharts/highcharts/pull/23803 (DraggablePoints  & Breadcrumbs)
- [x] Exporting, ExportData with [changes](https://github.com/highcharts/highcharts/pull/23683#issuecomment-3558967619) - reviewed

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206591292618188

KacperMadej commented:
Internal notes for the current review: 

- checking how .d.ts are build quick with `npm x gulp scripts-ts` that creates .d.ts files in `code/es-modules`
- StockTools & annotations are in the Stock interest zone, but I'm open to any review comments
- the current API and class-reference docs shouldn't change to worse or at all - make sure to build code before building the docs (so `npx gulp scripts --force` and then `npx gulp jsdoc`)
- side note: local class-reference files are not linked to end with .html and often they need it to open, so add `.html` in the URL manually in the browser when `Cannot GET ...` is showing before panicking.

kamil-musialowski commented on ts/Extensions/Annotations/Types/TimeCycles.ts:
```diff
@@ -337,6 +342,17 @@ interface TimeCycles {
 
 namespace TimeCycles {
+    /**
+     * Options for the  time cycles annotation type.
```
```suggestion
     * Options for the time cycles annotation type.
```

kamil-musialowski commented on ts/Extensions/Annotations/AnnotationDefaults.ts:
```diff
@@ -46,6 +46,12 @@ const { defined } = U;
+     *
+     * @type      {string}
+     * @apioption annotations.className
+     */
```
I think we could add `@since` param here, as I believe it will be available since next release and it wasn't available previously.

magdalena-gut reviewed (commented):
Review:
---

- [x] - means checked and ok
- [ ]  - menas checked but some improvements needed

---
Extensions/Annotations/Controllables
- [ ] /Controllable.ts 
- missing //internal on export default
- lack of doclets in /ControllableOptions.ts (ControllableLabelOptions, ControllableShapeOptions) 
- [x] Extensions/Annotations/Popup
- [ ] Extensions/Annotations/Types
- missing //internal on export default in most files
- missing  //internal  on the VerticalLine class declaration.
- [ ] Extensions/Annotations/
- /EventEmitter.ts - missing //internal on export default
- /NavigationBindings.ts - missing //internal on export default 
- [x] Extensions/Boost
- [ ] Extensions/Breadcrumbs
- /Breadcrumbs.ts - should class Breadcrumbs be internal?
- [ ] Extensions/DataGrouping
- /ApproximationDefaults.ts - missing //internal on the export default
- [x] Extensions/Debugger
- [x] Extensions/DraggablePoints
- [x] Extensions/DragPanes 
- [ ] Extensions/Drilldown
- Drilldown.ts - missing //internal tag for export default
- [ ] Extensions/ExportData
- ExportData.ts - function toggleDataTable has inorrect tag (line 1440: function Highcharts.Exporting#hideData)
- [ ] Extensions/Exporting
- ExportingBase.ts missing //internal tags?
- Fullscreen.ts - @intetnal tag move to the top for consitency 
- [ ] Extensions/MarkerClusters
- missing  //internal tags on export default for most files
- /MarkerClusters.ts - missing  //internal tags before declare module PointBase,
- /MarkerClusterDefaults.ts - missing  //internal tags before const MarkerClusterDefaults,
- /MarkerClusterScatter.ts - missing  //internal tags before const MarkerClusterScatter,
- /MarkerClusterSymbols.ts -  missing  //internal tags before const MarkerClusterSymbols,
- [ ] Extensions/MouseWheelZoom
- /MouseWheelZoom - missing //internal for MouseWheelZoomComposition 
- [ ] Extensions/NoDataDisplay
 - /NoDataToDisplay - missing //internal tag for export default
- /NoDataDefaults - missing //internal tag for export default
- [ ] Extensions/NonCartesianSeriesZoom
- /NonCartesianSeriesZoom.ts - 
- (line 57) - /* /* * should be /* *
- SeriesBase module missing //internal
- NonCartesianSeriesZoom class missing //internal
- export default missing //internal
- [ ] Extensions/OfflineExporting
- /OfflineExporting.ts - missing //internal for ChartBase and GlobalBase
- /OfflineExportingDefaults.ts - missing //internal for export
- [ ] Extensions/Pane
- /Pane.ts - missing //internal for AxisBase
- /PaneComposition.ts -  missing //internal for ChartBase; before PaneComposition; export default
- /PaneDefaults.ts  - missing //internal before export default
- [ ] Extensions/ParallelCoordinates
- /ParallelAxis.ts -  missing //internal for AxisComposition; and export default
- /ParallelCoordinates.ts - missing //internal for ChartBase; and export default
- /ParallelSeries.ts -  missing //internal for export default 
- /ParallelCoordinatesDefaults.ts -   missing //internal for export default 
- [ ] Extensions/SeriesLabel
- /SeriesLabel.ts -  missing //internal for ChartBase and SeriesBase; before const SeriesLabel;before export default
- /SeriesLabelDefaults.ts - missing //internal before export default
- /SeriesLabelUtilities.ts - missing //internal before export default
- [ ] Extensions/Sonification
- /Sonification.ts - missing //internal for declare module ChartBase; namespace Sonification
- /SonificationTimeline.ts - missing //internal for class SonificationTimeline; export default
- /TimelineChannel.ts - missing //internal for class TimelineChannel; export default
- /TimelineFromChart.ts - missing //internal for function timelineFromChart; export default 
- /MIDI.ts - missing //internal for function toMIDI; export default
- /Scales.ts - missing //internal for const Scales; export default Scales
- /InstrumentPresets.ts - missing //internal for const InstrumentPresets; export default
- /Options.ts - missing //internal for const Options; export default
- [ ] Extensions/Themes
- /Adaptive.ts - missing //internal for const defaultRules; const darkRules; const styleSheet
- /HighContrastDark.ts - missing //internal for const textBright
- [ ] Extensions/files
- /ArrowSymbols.ts - missing //internal for const ArrowSymbols; export default
- /BoostCanvas.ts - missing //internal for namespace BoostCanvas; export default; declare module SeriesBase
- /BorderRadius.ts - missing //internal for const BorderRadius; export default
- /CurrentDateIndication.ts - missing //internal for const CurrentDateIndication; export default
- /PatternFill.ts - missing //internal for const PatternFill; export default; declare module PointBase
- /PriceIndication.ts - missing //internal for const PriceIndication; export default; declare module SeriesBase
- /StaticScale.ts - missing //internal for const StaticScale; export default; declare module ChartBase

bre1470 reviewed (changes requested):
Great work so far. Requested changes:

- [ ] `ts/Extensions/DownloadURL.ts` has `@private` tags
- [ ] `ts/Extensions/Annotations/Controllables/ControllableOptions.ts` is not in sync (doclets, visibilities) with API options `annotations.labelOptions` and `annotations.shapeOptions`
- [ ] `ts/Extensions/Annotations/Types/*.ts` should not export anything afaik.
- [x] `ts/Extensions/Annotations/AnnotationDefaults.ts` should be internal.
- [x] `ts/Extensions/Annotations/AnnotationOptions.ts` should not export `AnnotationTypePointsOptions` and `ChartOptions`
- [ ] `ts/Extensions/Annotations/AnnotationSeries.ts` should be internal.
- [ ] In general anything `MockPoint` related should be internal as it is supposed to be transparent.
- [x] `ts/Extensions/Annotations/NavigationBindingsDefaults.ts` should be internal.
- [x] `ts/Extensions/Annotations/NavigationBindingsOptions.ts` lacks doclets for declared options.
- `ts/Extensions/Boost/**` looks good to me.
- [x] `ts/Extensions/Breadcrumbs/BreadcrumbsOptions.ts` lacks doclet in drilldown declaration.
- `ts/Extensions/DataGrouping/**` looks good to me.
- `ts/Extensions/Debugger/**` looks good to me.
- `ts/Extensions/Boost/**` looks good to me.
- [x] `ts/Extensions/DraggablePoints/DragDropDefaults.ts` should be internal.
- [x] `ts/Extensions/DraggablePoints/DragDropOptions.ts` should have doclets for `DragDropGuideBoxOptions` and not export `DragDropGuideBoxDefaultOptions`.
- [x] `ts/Extensions/DraggablePoints/DragDropProps.ts` lacks some doclets for `dragDrop`.
- `ts/Extensions/DragPanes/**` looks good to me. (noted a rename)
- [ ] `ts/Extensions/Drilldown/DrilldownDefaults.ts` should be internal.
- [x] `ts/Extensions/Drilldown/DrilldownOptions.ts` lacks defaults for `DrilldownActiveDataLabelStyleOptions`.
- `ts/Extensions/ExportData/**` looks good to me.
- [x] `ts/Extensions/Exporting/ExportingVendor.ts` should be internal.

KacperMadej commented:
Thank you @bre1470 for the review.

Some points were a bit unclear to me so here are some questions:
- there's no `ts/Extensions/DownloadURL.ts`. I changed `private` tags into `internal` in `ts\Shared\DownloadURL.ts` instead.
- Annotations code is casting to `any` making some types false. I changed interfaces to at least get the docs-related types to be correct - that was the intention. I checked briefly with the Stock Team if the proposed changes are not totally off, but a code review for that will be done too. For every added cast I tried to remove at least 1 other cast. Some of the related changes are also in line with another task of making options fully optional.
- "ts/Extensions/Annotations/Types/*.ts should not export anything afaik." - API options doclets in some files (advanced annotations for Stock) are defined there, so it should export it, right?
- "ts/Extensions/Annotations/AnnotationSeries.ts should be internal." - yes, but just added a needed empty interface, so now should be ok?
- "In general anything MockPoint related should be internal as it is supposed to be transparent." - Control points, formatter context, etc. - all those parts of the public API docs and are working with MockPoints. It doesn't look like the code was designed to keep MockPoint internal. I'm really trying to keep all that internal, but it leads to more casting or maybe some future refactoring. Internal options are mixed with API options at every usage level. I wanted to make this code better, but now I'm fine with not making it worse.
- "ts/Extensions/Drilldown/DrilldownDefaults.ts should be internal." - looks internal to me. Removing the doclet from the declared const DrilldownDefaults will break the JSDoc-based docs, right? or was this comment about something else?

Other points that I felt are now resolved are checked in your list.

bre1470 commented on ts/Extensions/Drilldown/DrilldownDefaults.ts:
```diff
@@ -33,11 +33,6 @@ import { Palette } from '../../Core/Color/Palettes.js';
- *
  * @sample {highcharts} highcharts/series-organization/drilldown
  *         Organization chart drilldown
  *
```
```
 * @product      highcharts highmaps
 * @requires     modules/drilldown
 * @optionparent drilldown
 * @internal
 */
const DrilldownDefaults: DrilldownOptions = {
```

bre1470 reviewed (changes requested):
`Shared/DownloadURL.d.ts` produces errors for me, because the `typeof` reflections do not find the references.
